[FX][testing] Test tracing into all the standard torch.nn.functional

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55550 [FX][testing] Test tracing into all the standard torch.nn.functional**

Add a test for `symbolic_trace` into `torch.nn.functional`

Test against all `functional`s with `torch.Tensor` argument and `functional`s from `FUNCTIONALS_WITHOUT_ANNOTATION`.
```py
FUNCTIONALS_WITHOUT_ANNOTATION = (
        "adaptive_max_pool1d",
        "adaptive_max_pool2d",
        "adaptive_max_pool3d",
        "fractional_max_pool2d",
        "fractional_max_pool3d",
        "max_pool1d",
        "max_pool2d",
        "max_pool3d",
        "gaussian_nll_loss",
        "upsample",
        "upsample_bilinear",
        "upsample_nearest",
    )
```

`UNTRACEABLE_FUNCTIONALS` lists 110 current untraceable `functional`s with expected `Error`.
- `BUILT_IN_FUNC`: built-in functions or built-in methods can not be traced.
- `PROXY_ITERABLE`: argument of type 'Proxy' is not iterable"
- `PROXY_ITERATED`: Proxy object cannot be iterated. This can be attempted when used in a for loop or as a *args or **kwargs function argument
- `LEN_ERROR`: 'len' is not supported in symbolic tracing by default. If you want this call to be recorded, please call torch.fx.wrap('len') at module scope
- `ARG_TYPE_MISMATCH`: `functional()`: argument <name> (position <n>) must be <type>, not Proxy
- `CONTROL_FLOW`: symbolically traced variables cannot be used as inputs to control flow
- `INTERPOLATE_ARGS_CONFLICT`: When tracing the functional by calling `interpolate(input, size, scale_factor, mode="bilinear", align_corners=True)`, `ValueError("only one of size or scale_factor should be defined")` is raised

## Note:
There is an inconsistent behavior due to different versions of Python. Use the following code as an example
```py
class P:
    def __iter__(self):
        raise RuntimeError

if __name__ == '__main__':
    a = P()
    if 1 in a:
        pass
```
- For Python 3.8, it raises `RuntimeError` as shown in the [python doc](https://docs.python.org/3.7/reference/expressions.html#membership-test-operations)
- For Python 3.6, it would raise `TypeError: argument of type 'P' is not iterable`
- Havn
- Haven't validated other Python versions, it seems python 3.8 is the only one can expose the internal error from `in` operation.

This inconsistent behavior will cause FX raise inconsistent Error from here:
https://github.com/pytorch/pytorch/blob/75eb026e07062e8539c3b6e96d6100ac9378a292/torch/_jit_internal.py#L395
When `kwargs` is a `Proxy` object, the `__iter__` function of `Proxy` instance would eventually call the `iter` function of `TraceBase` instance, which will raise Error.
https://github.com/pytorch/pytorch/blob/75eb026e07062e8539c3b6e96d6100ac9378a292/torch/fx/proxy.py#L131-L138

So, at the end, it will cause inconsistent behavior across different Python versions to `symbolic_trace` the functionals that use `boolean_dispatch `.

Differential Revision: [D27659367](https://our.internmc.facebook.com/intern/diff/D27659367)

ejguan committed:
[FX][testing] Test tracing into all the standard torch.nn.functional

[ghstack-poisoned]

jamesr66a reviewed (commented):
OK great! There's a couple things I think we need:

1) to disable `has_torch_function...` in `nn.functional`. Otherwise, we won't be tracing into the `nn.functional`. but rather just emitting a call to the functional directly.
2) Populate exception list.
3) Figure out how to only test tracing where Tensor valued inputs are traced. This could be done by using `inspect.signature` to figure out the type annotations of the Functional, then generating a function that only takes the `Tensor`-typed inputs, and calls into the functional with constant values for the other inputs.

I took a stab at (1) and (2) https://gist.github.com/jamesr66a/b6423f1d61c42b51464fc753b1289c04 feel free to incorporate that into the PR. Do you want to see about (3)?

ejguan commented:
> 1. to disable `has_torch_function...` in `nn.functional`. Otherwise, we won't be tracing into the `nn.functional`. but rather just emitting a call to the functional directly.

Ah, finally I understood this hack by replacing `has_torch_function ` with the dummy function always return False to ignore the __torch_function__ dispacth.

> 2. Populate exception list.

Working on that.

> 3. Figure out how to only test tracing where Tensor valued inputs are traced. This could be done by using `inspect.signature` to figure out the type annotations of the Functional, then generating a function that only takes the `Tensor`-typed inputs, and calls into the functional with constant values for the other inputs.

Will do. Pretty straight forward.

> I took a stab at (1) and (2) https://gist.github.com/jamesr66a/b6423f1d61c42b51464fc753b1289c04 feel free to incorporate that into the PR. Do you want to see about (3)?

Thanks for the patch, it helps me a lot.
And I will make each functional into a test under the TestFunctionalTracing. So, user can test each functional independently.

ejguan committed:
Update on "[FX][testing] Test tracing into all the standard torch.nn.functional"




[ghstack-poisoned]

ejguan commented on test/test_fx.py:
```diff
@@ -2258,6 +2259,218 @@ def test_get_torch_func_signature_exhaustive(self, device, dtype, op):
+            # Ignore non-callable object like modules
+            if not isinstance(fn, Callable):
+                continue
+            # Remove built-in function/method
+            #  if not hasattr(fn, '__module__') or fn.__module__ != 'torch.nn.functional':
+            #      continue
```
Like `adaptive_avg_pool1d`, it's bound to `torch.adaptive_avg_pool1d`, which is not a functional. We can ignore them can remove the `BUILT_IN_FUNC` from expected error.

jamesr66a replied:
Is this commented check still needed? Seems like this case is covered by BUILT_IN_FUNC exception checking

ejguan committed:
Update on "[FX][testing] Test tracing into all the standard torch.nn.functional"


Add a test for `symbolic_trace` into `torch.nn.functional`

Test against all `functional`s with `torch.Tensor` argument and `functional`s from `FUNCTIONALS_WITHOUT_ANNOTATION`.
```py
FUNCTIONALS_WITHOUT_ANNOTATION = (
        "adaptive_max_pool1d",
        "adaptive_max_pool2d",
        "adaptive_max_pool3d",
        "fractional_max_pool2d",
        "fractional_max_pool3d",
        "max_pool1d",
        "max_pool2d",
        "max_pool3d",
        "gaussian_nll_loss",
        "upsample",
        "upsample_bilinear",
        "upsample_nearest",
    )
```

`UNTRACEABLE_FUNCTIONALS` lists 110 current untraceable `functional`s with expected `Error`.
- `BUILT_IN_FUNC`: built-in functions or built-in methods can not be traced.
- `PROXY_ITERATED`: Proxy object cannot be iterated. This can be attempted when used in a for loop or as a *args or **kwargs function argument
- `LEN_ERROR`: 'len' is not supported in symbolic tracing by default. If you want this call to be recorded, please call torch.fx.wrap('len') at module scope
- `ARG_TYPE_MISMATCH`: `functional()`: argument <name> (position <n>) must be <type>, not Proxy
- `CONTROL_FLOW`: symbolically traced variables cannot be used as inputs to control flow
- `INTERPOLATE_ARGS_CONFLICT`: When tracing the functional by calling `interpolate(input, size, scale_factor, mode="bilinear", align_corners=True)`, `ValueError("only one of size or scale_factor should be defined")` is raised

Differential Revision: [D27659367](https://our.internmc.facebook.com/intern/diff/D27659367)

[ghstack-poisoned]

ejguan committed:
Update on "[FX][testing] Test tracing into all the standard torch.nn.functional"


Add a test for `symbolic_trace` into `torch.nn.functional`

Test against all `functional`s with `torch.Tensor` argument and `functional`s from `FUNCTIONALS_WITHOUT_ANNOTATION`.
```py
FUNCTIONALS_WITHOUT_ANNOTATION = (
        "adaptive_max_pool1d",
        "adaptive_max_pool2d",
        "adaptive_max_pool3d",
        "fractional_max_pool2d",
        "fractional_max_pool3d",
        "max_pool1d",
        "max_pool2d",
        "max_pool3d",
        "gaussian_nll_loss",
        "upsample",
        "upsample_bilinear",
        "upsample_nearest",
    )
```

`UNTRACEABLE_FUNCTIONALS` lists 110 current untraceable `functional`s with expected `Error`.
- `BUILT_IN_FUNC`: built-in functions or built-in methods can not be traced.
- `PROXY_ITERATED`: Proxy object cannot be iterated. This can be attempted when used in a for loop or as a *args or **kwargs function argument
- `LEN_ERROR`: 'len' is not supported in symbolic tracing by default. If you want this call to be recorded, please call torch.fx.wrap('len') at module scope
- `ARG_TYPE_MISMATCH`: `functional()`: argument <name> (position <n>) must be <type>, not Proxy
- `CONTROL_FLOW`: symbolically traced variables cannot be used as inputs to control flow
- `INTERPOLATE_ARGS_CONFLICT`: When tracing the functional by calling `interpolate(input, size, scale_factor, mode="bilinear", align_corners=True)`, `ValueError("only one of size or scale_factor should be defined")` is raised

Differential Revision: [D27659367](https://our.internmc.facebook.com/intern/diff/D27659367)

[ghstack-poisoned]

ejguan committed:
Update on "[FX][testing] Test tracing into all the standard torch.nn.functional"


Add a test for `symbolic_trace` into `torch.nn.functional`

Test against all `functional`s with `torch.Tensor` argument and `functional`s from `FUNCTIONALS_WITHOUT_ANNOTATION`.
```py
FUNCTIONALS_WITHOUT_ANNOTATION = (
        "adaptive_max_pool1d",
        "adaptive_max_pool2d",
        "adaptive_max_pool3d",
        "fractional_max_pool2d",
        "fractional_max_pool3d",
        "max_pool1d",
        "max_pool2d",
        "max_pool3d",
        "gaussian_nll_loss",
        "upsample",
        "upsample_bilinear",
        "upsample_nearest",
    )
```

`UNTRACEABLE_FUNCTIONALS` lists 110 current untraceable `functional`s with expected `Error`.
- `BUILT_IN_FUNC`: built-in functions or built-in methods can not be traced.
- `PROXY_ITERATED`: Proxy object cannot be iterated. This can be attempted when used in a for loop or as a *args or **kwargs function argument
- `LEN_ERROR`: 'len' is not supported in symbolic tracing by default. If you want this call to be recorded, please call torch.fx.wrap('len') at module scope
- `ARG_TYPE_MISMATCH`: `functional()`: argument <name> (position <n>) must be <type>, not Proxy
- `CONTROL_FLOW`: symbolically traced variables cannot be used as inputs to control flow
- `INTERPOLATE_ARGS_CONFLICT`: When tracing the functional by calling `interpolate(input, size, scale_factor, mode="bilinear", align_corners=True)`, `ValueError("only one of size or scale_factor should be defined")` is raised

Differential Revision: [D27659367](https://our.internmc.facebook.com/intern/diff/D27659367)

[ghstack-poisoned]

ejguan commented:
Hey @jamesr66a,
Spent some time on the CI error of pytorch_linux_bionic_py3_8_gcc9_coverage_test2. I find it caused by Python versions as I mentioned in the Note.
There are two options to solve it:
- Add different expected Exception for different python version
- Ignore the functionals that are implemented by `boolean_dispatch`

Personally, I think the first option makes more sense. But, I do want to hear your opinion.

ejguan committed:
Update on "[FX][testing] Test tracing into all the standard torch.nn.functional"


Add a test for `symbolic_trace` into `torch.nn.functional`

Test against all `functional`s with `torch.Tensor` argument and `functional`s from `FUNCTIONALS_WITHOUT_ANNOTATION`.
```py
FUNCTIONALS_WITHOUT_ANNOTATION = (
        "adaptive_max_pool1d",
        "adaptive_max_pool2d",
        "adaptive_max_pool3d",
        "fractional_max_pool2d",
        "fractional_max_pool3d",
        "max_pool1d",
        "max_pool2d",
        "max_pool3d",
        "gaussian_nll_loss",
        "upsample",
        "upsample_bilinear",
        "upsample_nearest",
    )
```

`UNTRACEABLE_FUNCTIONALS` lists 110 current untraceable `functional`s with expected `Error`.
- `BUILT_IN_FUNC`: built-in functions or built-in methods can not be traced.
- `PROXY_ITERABLE`: argument of type 'Proxy' is not iterable"
- `PROXY_ITERATED`: Proxy object cannot be iterated. This can be attempted when used in a for loop or as a *args or **kwargs function argument
- `LEN_ERROR`: 'len' is not supported in symbolic tracing by default. If you want this call to be recorded, please call torch.fx.wrap('len') at module scope
- `ARG_TYPE_MISMATCH`: `functional()`: argument <name> (position <n>) must be <type>, not Proxy
- `CONTROL_FLOW`: symbolically traced variables cannot be used as inputs to control flow
- `INTERPOLATE_ARGS_CONFLICT`: When tracing the functional by calling `interpolate(input, size, scale_factor, mode="bilinear", align_corners=True)`, `ValueError("only one of size or scale_factor should be defined")` is raised

## Note:
There is an inconsistent behavior due to different versions of Python. Use the following code as an example
```py
class P:
    def __iter__(self):
        raise RuntimeError

if __name__ == '__main__':
    a = P()
    if 1 in a:
        pass
```
- For Python 3.8, it raises `RuntimeError` as shown in the [python doc](https://docs.python.org/3.7/reference/expressions.html#membership-test-operations)
- For Python 3.6, it would raise `TypeError: argument of type 'P' is not iterable`
- Havn
- Haven't validated other Python versions, it seems python 3.8 is the only one can expose the internal error from `in` operation.

This inconsistent behavior will cause FX raise inconsistent Error from here:
https://github.com/pytorch/pytorch/blob/75eb026e07062e8539c3b6e96d6100ac9378a292/torch/_jit_internal.py#L395
When `kwargs` is a `Proxy` object, the `__iter__` function of `Proxy` instance would eventually call the `iter` function of `TraceBase` instance, which will raise Error.
https://github.com/pytorch/pytorch/blob/75eb026e07062e8539c3b6e96d6100ac9378a292/torch/fx/proxy.py#L131-L138

So, at the end, it will cause inconsistent behavior across different Python versions to `symbolic_trace` the functionals that use `boolean_dispatch `.

Differential Revision: [D27659367](https://our.internmc.facebook.com/intern/diff/D27659367)

[ghstack-poisoned]

ejguan commented on test/test_fx.py:
```diff
@@ -2258,6 +2259,251 @@ def test_get_torch_func_signature_exhaustive(self, device, dtype, op):
+        "upsample_nearest",
+    )
+
+    # For the operator like `in`, internal exception during the iteration
+    # can be raised on Python 3.8, rather than the error of `object is not iterable`
+    UNTRACEABLE_FUNCTIONALS_PY38 = {
+        "adaptive_max_pool1d": PROXY_ITERATED,
+        "adaptive_max_pool2d": PROXY_ITERATED,
+        "adaptive_max_pool3d": PROXY_ITERATED,
+        "fractional_max_pool2d": PROXY_ITERATED,
+        "fractional_max_pool3d": PROXY_ITERATED,
+        "max_pool1d": PROXY_ITERATED,
+        "max_pool2d": PROXY_ITERATED,
+        "max_pool3d": PROXY_ITERATED,
+
+        "group_norm": LEN_ERROR
+    }
```
Has to implement this hack

jamesr66a replied:
I don't 100% get the context here, can you paste a stack trace or something?

It's great that we're uncovering issues like this BTW, thanks!

ejguan replied:
Use `adaptive_max_pool1d ` as an example. When we do `symbolic_trace` over it, the Error is different with different Python version.
- Python 3.8
```py
```
- Other (Use Python 3.6 as an example)
```py
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/erjia/workspace/pytorch-dev/py36/torch/fx/symbolic_trace.py", line 674, in symbolic_trace
    graph = tracer.trace(root, concrete_args)
  File "/Users/erjia/workspace/pytorch-dev/py36/torch/fx/symbolic_trace.py", line 419, in trace
    self.create_node('output', 'output', (self.create_arg(fn(*args)),), {},
  File "/Users/erjia/workspace/pytorch-dev/py36/torch/_jit_internal.py", line 395, in fn
    if arg_name in kwargs:
TypeError: argument of type 'Proxy' is not iterable
```

The error is caused here when `kwargs` is a Proxy object
https://github.com/pytorch/pytorch/blob/75eb026e07062e8539c3b6e96d6100ac9378a292/torch/_jit_internal.py#L395
When Proxy object is operated by `in`, `__iter__` function will be called and it will eventually raise a TraceError from the `iter` function of `TraceBase`. But, it's able to re-raise the internal TraceError with Python 3.8. It will raise `<object> is not iterable` with other python versions.

You can use the example in the note of main comment to verify this behavior.

And, I will update the comment here to make it easier to understand.

jamesr66a commented on test/test_fx.py:
```diff
@@ -2258,6 +2259,251 @@ def test_get_torch_func_signature_exhaustive(self, device, dtype, op):
+                'has_torch_function_unary': None,
+                'has_torch_function_variadic': None}
+
+    BUILT_IN_FUNC = (AssertionError, '')
```
Is there no exception message for htis one?

ejguan replied:
Yes

jamesr66a commented on test/test_fx.py:
```diff
@@ -2258,6 +2259,251 @@ def test_get_torch_func_signature_exhaustive(self, device, dtype, op):
+                    'boolean_dispatch')
+    TO_PATCH = {'has_torch_function': None,
+                'has_torch_function_unary': None,
+                'has_torch_function_variadic': None}
```
nit: the rest of the strings in this test use "double quotes"

jamesr66a reviewed (approved):
Great, thanks so much for doing this! I mostly just had questions in the review comments, otherwise looks like this is good to go!

ejguan committed:
Update on "[FX][testing] Test tracing into all the standard torch.nn.functional"


Add a test for `symbolic_trace` into `torch.nn.functional`

Test against all `functional`s with `torch.Tensor` argument and `functional`s from `FUNCTIONALS_WITHOUT_ANNOTATION`.
```py
FUNCTIONALS_WITHOUT_ANNOTATION = (
        "adaptive_max_pool1d",
        "adaptive_max_pool2d",
        "adaptive_max_pool3d",
        "fractional_max_pool2d",
        "fractional_max_pool3d",
        "max_pool1d",
        "max_pool2d",
        "max_pool3d",
        "gaussian_nll_loss",
        "upsample",
        "upsample_bilinear",
        "upsample_nearest",
    )
```

`UNTRACEABLE_FUNCTIONALS` lists 110 current untraceable `functional`s with expected `Error`.
- `BUILT_IN_FUNC`: built-in functions or built-in methods can not be traced.
- `PROXY_ITERABLE`: argument of type 'Proxy' is not iterable"
- `PROXY_ITERATED`: Proxy object cannot be iterated. This can be attempted when used in a for loop or as a *args or **kwargs function argument
- `LEN_ERROR`: 'len' is not supported in symbolic tracing by default. If you want this call to be recorded, please call torch.fx.wrap('len') at module scope
- `ARG_TYPE_MISMATCH`: `functional()`: argument <name> (position <n>) must be <type>, not Proxy
- `CONTROL_FLOW`: symbolically traced variables cannot be used as inputs to control flow
- `INTERPOLATE_ARGS_CONFLICT`: When tracing the functional by calling `interpolate(input, size, scale_factor, mode="bilinear", align_corners=True)`, `ValueError("only one of size or scale_factor should be defined")` is raised

## Note:
There is an inconsistent behavior due to different versions of Python. Use the following code as an example
```py
class P:
    def __iter__(self):
        raise RuntimeError

if __name__ == '__main__':
    a = P()
    if 1 in a:
        pass
```
- For Python 3.8, it raises `RuntimeError` as shown in the [python doc](https://docs.python.org/3.7/reference/expressions.html#membership-test-operations)
- For Python 3.6, it would raise `TypeError: argument of type 'P' is not iterable`
- Havn
- Haven't validated other Python versions, it seems python 3.8 is the only one can expose the internal error from `in` operation.

This inconsistent behavior will cause FX raise inconsistent Error from here:
https://github.com/pytorch/pytorch/blob/75eb026e07062e8539c3b6e96d6100ac9378a292/torch/_jit_internal.py#L395
When `kwargs` is a `Proxy` object, the `__iter__` function of `Proxy` instance would eventually call the `iter` function of `TraceBase` instance, which will raise Error.
https://github.com/pytorch/pytorch/blob/75eb026e07062e8539c3b6e96d6100ac9378a292/torch/fx/proxy.py#L131-L138

So, at the end, it will cause inconsistent behavior across different Python versions to `symbolic_trace` the functionals that use `boolean_dispatch `.

Differential Revision: [D27659367](https://our.internmc.facebook.com/intern/diff/D27659367)

[ghstack-poisoned]

ejguan commented:
@jamesr66a 
Your suggestions are addressed and update the comment for `UNTRACEABLE_FUNCTIONALS_PY38`.
